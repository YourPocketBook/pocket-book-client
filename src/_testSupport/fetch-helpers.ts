import fetchMock, {
  MockOptionsMethodDelete,
  MockOptionsMethodGet,
  MockOptionsMethodPost,
  MockOptionsMethodPut,
  MockResponse,
  MockResponseFunction
} from "fetch-mock";

afterEach(fetchMock.reset);

export function authenticatedGet(
  url: string,
  token: string,
  response: MockResponse | MockResponseFunction,
  options?: MockOptionsMethodGet
) {
  fetchMock.get(
    (u, h) =>
      u === url &&
      (h.headers as { Authorization: string }).Authorization ===
        `Bearer ${token}`,
    response,
    options
  );
}

export function authenticatedPut(
  url: string,
  token: string,
  body: {},
  response: MockResponse | MockResponseFunction,
  options?: MockOptionsMethodPut
) {
  fetchMock.put(
    (u, h) =>
      u === url &&
      (h.headers as { Authorization: string }).Authorization ===
        `Bearer ${token}` &&
      h.body === JSON.stringify(body),
    response,
    options
  );
}

export function authenticatedPost(
  url: string,
  token: string,
  body: {},
  response: MockResponse | MockResponseFunction,
  options?: MockOptionsMethodPost
) {
  fetchMock.post(
    (u, h) =>
      u === url &&
      (h.headers as { Authorization: string }).Authorization ===
        `Bearer ${token}` &&
      h.body === JSON.stringify(body),
    response,
    options
  );
}

export function authenticatedDelete(
  url: string,
  token: string,
  response: MockResponse | MockResponseFunction,
  options?: MockOptionsMethodDelete
) {
  fetchMock.delete(
    (u, h) =>
      u === url &&
      (h.headers as { Authorization: string }).Authorization ===
        `Bearer ${token}`,
    response,
    options
  );
}
