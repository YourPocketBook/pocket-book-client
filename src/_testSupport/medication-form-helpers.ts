import { fireEvent } from "@testing-library/react";

export const testDefaultState = {
  adviceIfDeclined: "Test Advice 1",
  adviceIfTaken: "Test Advice 2",
  dose: "Test Dose",
  exclusionCriteria: "Test Exclusion",
  form: "Test Form",
  inclusionCriteria: "Test Inclusion",
  indications: "Test Indications",
  name: "Test Name",
  policyDate: "2019-12-10",
  route: "Test Route",
  sideEffects: "Test Side Effects"
};

export function setupState(getByLabelText: (text: string) => HTMLElement) {
  fireEvent.change(getByLabelText("Name"), {
    target: { value: testDefaultState.name }
  });
  fireEvent.change(getByLabelText("Policy Date"), {
    target: { value: testDefaultState.policyDate }
  });
  fireEvent.change(getByLabelText("Indications"), {
    target: { value: testDefaultState.indications }
  });
  fireEvent.change(getByLabelText("Inclusion Criteria"), {
    target: { value: testDefaultState.inclusionCriteria }
  });
  fireEvent.change(getByLabelText("Exclusion Criteria"), {
    target: { value: testDefaultState.exclusionCriteria }
  });
  fireEvent.change(getByLabelText("Name, Forms and Strengths of Medicine"), {
    target: { value: testDefaultState.form }
  });
  fireEvent.change(getByLabelText("Route or Method"), {
    target: { value: testDefaultState.route }
  });
  fireEvent.change(getByLabelText("Dosage"), {
    target: { value: testDefaultState.dose }
  });
  fireEvent.change(getByLabelText("Side Effects"), {
    target: { value: testDefaultState.sideEffects }
  });
  fireEvent.change(getByLabelText("Advice if Taken"), {
    target: { value: testDefaultState.adviceIfTaken }
  });
  fireEvent.change(getByLabelText("Advice if Declined"), {
    target: { value: testDefaultState.adviceIfDeclined }
  });
}

export function checkForClass(
  getByLabelText: (text: string) => HTMLElement,
  cls: "is-valid" | "is-invalid"
) {
  expect(getByLabelText("Name")).toHaveClass(cls);
  expect(getByLabelText("Policy Date")).toHaveClass(cls);
  expect(getByLabelText("Indications")).toHaveClass(cls);
  expect(getByLabelText("Inclusion Criteria")).toHaveClass(cls);
  expect(getByLabelText("Exclusion Criteria")).toHaveClass(cls);
  expect(getByLabelText("Name, Forms and Strengths of Medicine")).toHaveClass(
    cls
  );
  expect(getByLabelText("Route or Method")).toHaveClass(cls);
  expect(getByLabelText("Dosage")).toHaveClass(cls);
  expect(getByLabelText("Side Effects")).toHaveClass(cls);
  expect(getByLabelText("Advice if Taken")).toHaveClass(cls);
  expect(getByLabelText("Advice if Declined")).toHaveClass(cls);
}

export function checkState(
  getByLabelText: (text: string) => HTMLElement,
  hasValue: boolean
) {
  expect(getByLabelText("Name")).toHaveValue(
    hasValue ? testDefaultState.name : ""
  );
  expect(getByLabelText("Policy Date")).toHaveValue(
    hasValue ? testDefaultState.policyDate : ""
  );
  expect(getByLabelText("Indications")).toHaveTextContent(
    hasValue ? testDefaultState.indications : ""
  );
  expect(getByLabelText("Inclusion Criteria")).toHaveTextContent(
    hasValue ? testDefaultState.inclusionCriteria : ""
  );
  expect(getByLabelText("Exclusion Criteria")).toHaveTextContent(
    hasValue ? testDefaultState.exclusionCriteria : ""
  );
  expect(
    getByLabelText("Name, Forms and Strengths of Medicine")
  ).toHaveTextContent(hasValue ? testDefaultState.form : "");
  expect(getByLabelText("Route or Method")).toHaveTextContent(
    hasValue ? testDefaultState.route : ""
  );
  expect(getByLabelText("Dosage")).toHaveTextContent(
    hasValue ? testDefaultState.dose : ""
  );
  expect(getByLabelText("Side Effects")).toHaveTextContent(
    hasValue ? testDefaultState.sideEffects : ""
  );
  expect(getByLabelText("Advice if Taken")).toHaveTextContent(
    hasValue ? testDefaultState.adviceIfTaken : ""
  );
  expect(getByLabelText("Advice if Declined")).toHaveTextContent(
    hasValue ? testDefaultState.adviceIfDeclined : ""
  );
}
