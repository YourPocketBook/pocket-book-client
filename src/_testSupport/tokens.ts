export const inDateNonAdminToken =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwiaWF0IjoxNDE2MjM5MDIyLCJleHAiOjE3MTYyNDAwMjIsImF1ZCI6ImxvY2FsaG9zdCIsImlzcyI6ImxvY2FsaG9zdCIsImdpdmVuX25hbWUiOiJUZXN0IFVzZXIiLCJuYW1laWQiOiJmZjFkMDhlNy00NjU2LTRjMjgtOTc2Zi0xYzI3ZjE5MTQ3N2YiLCJlbWFpbCI6InRlc3RAc2phLm9yZy51ayJ9.VmT5YkTG0sUd5sK7skfMdi9kc5XG7Yhx6sqhqMSxhVY";
export const inDateAdminToken =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwiaWF0IjoxNDE2MjM5MDIyLCJleHAiOjE3MTYyNDAwMjIsImF1ZCI6ImxvY2FsaG9zdCIsImlzcyI6ImxvY2FsaG9zdCIsInJvbGUiOiJBZG1pbiIsImdpdmVuX25hbWUiOiJUZXN0IFVzZXIiLCJuYW1laWQiOiJmZjFkMDhlNy00NjU2LTRjMjgtOTc2Zi0xYzI3ZjE5MTQ3N2YiLCJlbWFpbCI6InRlc3RAc2phLm9yZy51ayJ9.rSQb2Hv6RfxaXSfO6s3QjLjnYKeqVgps3rPenquh3Nk";
export const expiredToken =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwiaWF0IjoxNDE2MjM5MDIyLCJleHAiOjE0MTYyNDAwMjIsImF1ZCI6ImxvY2FsaG9zdCIsImlzcyI6ImxvY2FsaG9zdCIsInVuaXF1ZV9uYW1lIjoiVGVzdCBVc2VyIiwibmFtZWlkIjoiZmYxZDA4ZTctNDY1Ni00YzI4LTk3NmYtMWMyN2YxOTE0NzdmIiwiZW1haWwiOiJ0ZXN0QHNqYS5vcmcudWsifQ.7bDfHDk9sWXL8boFeZWIdo_GB-4Qj8PvA4HDC49EzNc";
