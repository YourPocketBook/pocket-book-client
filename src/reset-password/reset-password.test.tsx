import "@testing-library/jest-dom/extend-expect";
import { fireEvent, render, wait } from "@testing-library/react";
import deepEqual from "deep-equal";
import fetchMock from "fetch-mock";
import React from "react";
import ReactDOM from "react-dom";
import { ResetPassword } from ".";
import "../_testSupport/fetch-helpers";
import { history } from "../history";

jest.mock("zxcvbn");

const endpoint = "/api/reset-password";

const testDefaultState = {
  password: "testPassword"
};

describe("Reset Password Page", () => {
  beforeEach(() => {
    history.push = jest.fn();
    history.replace = jest.fn();
  });

  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<ResetPassword userId="1234" code="4567" />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it("renders correctly", () => {
    const { container } = render(<ResetPassword userId="1234" code="4567" />);

    expect(container).toMatchSnapshot();
  });

  it("defaults to correct values", () => {
    const { getByLabelText } = render(
      <ResetPassword userId="1234" code="4567" />
    );

    expect(getByLabelText("New Password")).toHaveAttribute("value", "");
    expect(getByLabelText("New Password")).toHaveAttribute("type", "password");
  });

  it("flags valid values correctly", async () => {
    fetchMock.post(
      (u, h) =>
        u === endpoint &&
        deepEqual(JSON.parse((h.body as string) || ""), {
          code: "4567",
          newPassword: testDefaultState.password,
          userId: "1234"
        }),
      {
        body: JSON.stringify([{ path: "doses", error: "test" }]),
        status: 400
      }
    );

    const { getByLabelText, getByText } = render(
      <ResetPassword userId="1234" code="4567" />
    );

    fireEvent.change(getByLabelText("New Password"), {
      target: { value: testDefaultState.password }
    });
    fireEvent.click(getByText("Submit"));

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(getByLabelText("New Password")).toHaveClass("is-valid");
  });

  it("flags invalid values correctly", () => {
    const { getByLabelText, getByText } = render(
      <ResetPassword userId="1234" code="4567" />
    );

    fireEvent.change(getByLabelText("New Password"), {
      target: { value: "" }
    });
    fireEvent.click(getByText("Submit"));

    expect(getByLabelText("New Password")).toHaveClass("is-invalid");
  });

  it("renders server error correctly", async () => {
    fetchMock.post(
      (u, h) =>
        u === endpoint &&
        deepEqual(JSON.parse((h.body as string) || ""), {
          code: "4567",
          newPassword: testDefaultState.password,
          userId: "1234"
        }),
      {
        status: 500
      }
    );

    const { getByLabelText, getByText } = render(
      <ResetPassword userId="1234" code="4567" />
    );

    fireEvent.change(getByLabelText("New Password"), {
      target: { value: testDefaultState.password }
    });
    fireEvent.click(getByText("Submit"));

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(getByText(/The server has had an issue/)).toBeInTheDocument();
  });

  it("renders in progress correctly", async () => {
    fetchMock.post(
      (u, h) =>
        u === endpoint &&
        deepEqual(JSON.parse((h.body as string) || ""), {
          code: "4567",
          newPassword: testDefaultState.password,
          userId: "1234"
        }),
      // tslint:disable-next-line: no-empty
      new Promise(() => {})
    );

    const { getByLabelText, getByText } = render(
      <ResetPassword userId="1234" code="4567" />
    );

    fireEvent.change(getByLabelText("New Password"), {
      target: { value: testDefaultState.password }
    });
    fireEvent.click(getByText("Submit"));

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(getByText("Submit")).toBeDisabled();
  });

  it("toggles show password when button is clicked", () => {
    const { getByText, getByLabelText } = render(
      <ResetPassword userId="1234" code="4567" />
    );

    fireEvent.click(getByText("Show Password"));
    expect(getByLabelText("New Password")).toHaveAttribute("type", "text");

    fireEvent.click(getByText("Hide Password"));
    expect(getByLabelText("New Password")).toHaveAttribute("type", "password");
  });

  it("redirects on successful form submit", async () => {
    fetchMock.post(
      (u, h) =>
        u === endpoint &&
        deepEqual(JSON.parse((h.body as string) || ""), {
          code: "4567",
          newPassword: testDefaultState.password,
          userId: "1234"
        }),
      { status: 204 }
    );

    const { getByLabelText, getByText } = render(
      <ResetPassword userId="1234" code="4567" />
    );

    fireEvent.change(getByLabelText("New Password"), {
      target: { value: testDefaultState.password }
    });
    fireEvent.click(getByText("Submit"));

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(history.push).toBeCalledWith("/login");
  });

  it("redirects home when no userId", async () => {
    render(<ResetPassword userId="" code="4567" />);

    expect(history.replace).toBeCalledWith("/");
  });

  it("redirects home when no code", async () => {
    render(<ResetPassword userId="1234" code="" />);

    expect(history.replace).toBeCalledWith("/");
  });
});
