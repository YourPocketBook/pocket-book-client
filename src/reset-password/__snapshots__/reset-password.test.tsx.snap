// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reset Password Page renders correctly 1`] = `
<div>
  <div
    class="background"
  >
    <nav
      class="navbar navbar"
    >
      <a
        class="navbar-brand"
      >
        PocketBook
      </a>
    </nav>
    <div
      class="centreContainer container"
    >
      <div
        class="row"
      >
        <div
          class="col-lg-6 offset-lg-3"
        >
          <div
            class="centreBox"
          >
            <div
              class="container"
            >
              <div
                class="row"
              >
                <div
                  class="col"
                >
                  <h1
                    class="text-center"
                  >
                    Reset Your Password
                  </h1>
                </div>
              </div>
              <div
                class="row"
              >
                <div
                  class="col"
                >
                  <p>
                    Now we have confirmed your email address, you need to set a new password.
                  </p>
                </div>
              </div>
              <form
                class=""
                id="reset-password-form"
                novalidate=""
              >
                <div
                  class="row"
                >
                  <div
                    class="col"
                  >
                    <div
                      class="form-group"
                    >
                      <label
                        class=""
                        for="newPassword"
                      >
                        New Password
                      </label>
                      <input
                        autocomplete="new-password"
                        class="form-control"
                        id="newPassword"
                        name="newPassword"
                        placeholder="New Password"
                        type="password"
                        value=""
                      />
                      <div
                        class="invalid-feedback"
                      >
                        You need to enter a password.
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="row"
                >
                  <div
                    class="text-right col-md-9 offset-md-3"
                  >
                    <button
                      class="text-muted btn btn-link"
                      id="showPassword"
                      type="button"
                    >
                      Show Password
                    </button>
                  </div>
                </div>
                <button
                  class="btn btn-dark-green btn-block"
                  type="submit"
                >
                  Submit
                </button>
              </form>
            </div>
          </div>
        </div>
      </div>
    </div>
    <nav
      class="fixed-bottom footer navbar"
    >
      <span
        class="navbar-text"
      >
        Design Copyright Â© 
        2020
         Tony Richards.
      </span>
      <ul
        class="nav"
      >
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/privacy"
          >
            Privacy Policy
          </a>
        </li>
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/tsandcs"
          >
            Terms of Use
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;
